rfcs/20200525-elasticity-and-fault-tolerance-in-multiworkermirroredstrategy.md

```diff
@@ -0,0 +1,439 @@
+
+
+### **_CAVEAT: Model Determinism_**
+
+In the event of membership changes, to accommodate the new cluster size, the dataset will have to be re-distributed among the new set of working nodes. This will cause the model to lose determinism while training the model.
+Prefetching datasets
+Any stateful variables that are not created under the MWMS scope, will not be restored. All pertinent variables needed have to be created under the MWMS scope.
+
+
+### _Workflow_
+
+![Elastic Training Flow](./20200525-elasticity-and-fault-tolerance-in-multiworkermirroredstrategy/elastic_training_flow.png)
+
+
+
+The workflow shown above goes through major and typical situations during elastic training. This occurs within the strategy run function.
+
+* Normal process where the state is saved, but the function runs correctly. 
+* A membership change process that adds a new node (W2, W3). 
+* A communications failure where NetworkExceptions are raised, which initiates a membership change process to remove failing node. 
+* A node failure which causes a node to fail, and propagate by causing the system a network failure on the next iteration. 
+* A normal termination process by OutOfRangeException
+
+### _Variable Sync Process_
+
+The variable sync process is where the chief broadcasts its variables to the other workers to keep them in sync at the beginning. On a new cluster, each node will create their variables locally, then they will synchronize their variables with the chief. The Variable sync process can also be started when there is a new node being added to the cluster. This new node will need to synchronize its variables with the chief. Currently the chief will broadcast to everyone and only the new nodes will actually modify their variables. 
+
+* Broadcast from chief
+    * Catch NetworkError and start Membership Change Process
+* Assign variables from broadcast
+
+**Datasets**
+
+When a new process joins the run call, the dataset of the new nodes and the old nodes will be out of sync. Each dataset operator internally has an iterator object which can be restored through MakeIteratorFromCheckpoint. The chief will send its serialized iterator to be restored on the new nodes. This will bring the new nodes to the last iterated state of the dataset. 
+
+
+### _Membership Change Process_
+
+The membership change process happens when the cluster needs to be modified. This will take place when new nodes are being added to the cluster, or there is a NetworkException raised. The membership change process will attempt to figure out failed nodes, new nodes, and send out cluster spec updates to each worker. 
+
+* Everyone ping chief
+    * timeout for chief to respond
+    * If chief doesn’t respond ping next chief candidate
```

Comment:
It seems difficult for all workers to agree on the failure state of the chief without any coordination.

Reply:
That’s also my concern. We probably need a consensus algorithm, e.g. Paxos or Raft, if we wanted to do leader election in a proper way.

Reply:
> Thank you for your RFC! Do you think your proposal can be a layer above the existing MWMS, so that we can experiment with it and allow users to use it first while we think of ways to let it graduate it into TensorFlow core? I guess this is possible since your `discovery_script` is not coupled with any of TF's components. You can also extend MWMS for dataset-related changes.

Thank you!
Is the idea a separate repository or just a separate Strategy? This current idea has Core changes implied. A separate repo can be done once the graph rebuilding costs are understood.

Reply:
I think most of the changes can be in a separate repository (we've been thinking of creating a new repo). It is reasonable if some changes are needed in the TF core. But you can start with a prototype or a demo using monkey-patch, linking new kernels etc.

Reply:
Re consensus algorithm, I am not sure whether it is a good an idea to require users to configure these services. Another option is similar to [the single-client PS design](https://github.com/yuefengz/community/blob/master/rfcs/20200306-single-client-parameter-server.md) which uses a single coordinator for the whole training cluster. But it assumes the client won't fail, otherwise distributed consensus is still needed. These elastic resources (e.g. elastic dataset) may be better fit in the client layer (we'll separate the client and the strategy class).

Reply:
So this was actually a design that was considered. Using an "orchestrator". If the orchestrator fails, you will lose all elasticity and possibly fault tolerance. Another reason why I didn't pursue this was it will change the idea of MWMS. But it seems because adding elasticity is a huge feature, this will be encapsulated in its own strategy, So this can be introduced again.

Reply:
Adding some more data. 
Variable initialization is throughout the whole process This is Dataset and Function Tracing.
From my understanding of eager, each function is traced twice, one with Variable creation, and another without.
Any more models i should get data for?
| Strategy | Model | setup | Function Tracing (sec) | Variable Initialization (sec) | Dataset Initialization (sec) |
|----------|:-----:|:------:|:------:|:------:|:------:|
| multi-worker | efficientnet| 2 g4dn 4 gpus| 36.54 | 6.77 | 6.06 |
| multi-worker | resnet| 2 g4dn 4 gpus| 20.35 | 4.89 | 1.43 |
| multi-worker | transformer| 2 g4dn 4 gpus| 38.08 | 14.07 | 0.08 |
| multi-worker | bert| 2 g4dn 4 gpus| 55.32 | 24.5 | 0.28 |
| multi-worker | albert| 2 g4dn 4 gpus| 38.08 | 14.07 | 0.08 |